### ISSUE TITLE
[google_maps_flutter_web] An example to update a marker position

### ISSUE THREAD

**Issue by celsoncustodio** [flutter/flutter#75778]
How I update a marker's position, does anyone have an example? I've tried in several ways

**Comment by GAUTAMMISTRY** [comment]
Checkout here
https://github.com/flutter/flutter/issues/65255

**Comment by darshankawar** [comment]
@celsoncustodio 
Check above comment and see if it answers your question.

**Comment by celsoncustodio** [comment]
> Checkout here
> #65255
> & example is https://dit-maps-tests.web.app/#/

I did not understand.. i have try use setstate, but have not sucess, do u have a simple code example

**Comment by darshankawar** [comment]
@celsoncustodio 
Check out [this](https://github.com/flutter/flutter/issues/54515) issue and [this](https://github.com/flutter/flutter/issues/54515#issuecomment-673369050) solution and see if it helps.

**Comment by neoacevedo** [comment]
But such example is with List objectand Google Map accepts as marker parameter only Set object.

**Comment by Maverick099** [comment]
@neoacevedo  Originally when you  change the markers list and then call set state that should update the markers automatically because `setState(){}` while updating the app states sees that the original list data has been changed and in turn, this should update the `GoogleMap` widget.  but for some reason it is not happening, that is the reason I explicitly called the implementation functions.

Now I think there is no need of calling the functions explicitly, pass a key to the google map widget and then call `setState((){});` after updating the marker list 
pass it to the map by using `Set<Marker>.from(myMarkers)`

this should work now, because now `didUpdateWidget(oldWidget)` is overridden in the new version of the plugin and this should update the markers position without need of explicitly calling the implementation methods.

**Comment by celsoncustodio** [comment]
@darshankawar @Maverick099 
didn't work yet 

i'm using in flutter web, i think maybe is different then mobile, because of some bug

obs: in mobile i can update marker position normaly.

` setState(() {
          var setmarkersnew= markersnew.toSet();

          MarkerUpdates.from(
              Set<Marker>.from(setmarkers), Set<Marker>.from(setmarkersnew));
          setmarkers = [] as Set<Marker>;
          setmarkers = setmarkersnew;
        });`

**Comment by darshankawar** [comment]
Cc: @ditman

**Comment by ditman** [comment]
Thanks for bringing this to my attention. I'm going to close #65255 and use this one as tracking for both the issue of markers not updating correctly, and making sure the example is understandable.

I'll revisit this soon, once the first wave of web plugins migrated to null-safety lands!

**Comment by neoacevedo** [comment]
@Maverick099 Good, I created another project, only with google maps, Location package and a marker, without navigation, no timers or any other widget needed to be updated via setState, and it works good; in my current and private project I navigate from one screen to the next one where I have 1 timer firing around each 10 secs, the Location package and may widgets including google maps and the marker I want to update its location. Inside the location listener I try to update the marker location when the location changes and here is when I get the map flickering.

The code where I update the marker is inside the onMapCreated method:
```dart
  void _onMapCreated(GoogleMapController controller) {
    mapController = controller;
    _locationSubscription =
        location.onLocationChanged.listen((currentLocation) {
      // Controlar que con cualquier salto del GPS, no se mueva si no se está en movimiento.
      if (currentLocation.accuracy <= 8 && currentLocation.speed >= 1) {
        _center = LatLng(currentLocation.latitude, currentLocation.longitude);
        if (mapController != null) {
          markerDriverLocation =
              markerDriverLocation.copyWith(positionParam: _center);

          // Centrar el mapa
          mapController.animateCamera(CameraUpdate.newCameraPosition(
              CameraPosition(
                  target: LatLng(_center.latitude, _center.longitude),
                  zoom: zoom)));
          setState(() {
            _markers[markerDriverId] = markerDriverLocation;
          });
        }
      }
    });
  }
```

The screen has more or less the following structure:
```
WillPop
- MaterialApp
-- Scaffold
--- AppBar
--- SizedBox
---- Column
----- Container
------ Widget1
----- Container
------ Container
------- GoogleMaps
```

**Comment by celsoncustodio** [comment]
I did it this way, but the markers are flashing, someone would have another alternative ?

![image](https://user-images.githubusercontent.com/30812287/110783715-8e806480-8247-11eb-83fb-3aacca11d6e6.png)

**Comment by neoacevedo** [comment]
In a very basic example, the map never flicks, for instance: https://github.com/flutter/flutter/issues/43785#issuecomment-790855719 
But in a very more complex implementation, as I described in a previous comment, the map flicks on each location update.

**Comment by ditman** [comment]
I'm migrating the google_maps_flutter_web package to null-safety, and the latest version of the example app seems to work as expected; in the "Place marker" demo, after clicking "add" to add a marker, then selecting the marker and clicking "toggle draggable", the marker can be moved around.

The UI of that demo is not the best, but it seems that moving the marker is working?

The refreshing of the map, might be this [issue issue related to platform views in general](https://github.com/flutter/flutter/issues/51865).

(A video recording of what you're seeing might be useful to determine if the issue is the same or not.)

**Comment by neoacevedo** [comment]
@ditman does it applies to for google_maps_flutter for mobile too?

**Comment by ditman** [comment]
@neoacevedo the mobile plugin shouldn't have a flickering issue (or at least, not the same flickering issue as the Web version). Are you running into this in a native Android App, or on a mobile web browser, or on a desktop web browser?

**Comment by neoacevedo** [comment]
I'm running it in a native Android device. I also ran it into an emulator and I got the same results.

**Comment by ditman** [comment]
> I'm running it in a native Android app device. I also ran it into an emulator and I got the same results.

@neoacevedo Can you please file a new issue for this? If your problem is happening in Android, you're not using the `google_maps_flutter_web` package at all.

**Comment by kjawadDeveloper** [comment]
Does anybody found any solution? I am also experiencing this on the mobile side too.

**Comment by artificerchris** [comment]
> I'm migrating the google_maps_flutter_web package to null-safety

Is there an issue I can watch for this? I'm currently in dependency hell on a null safety migration because of that package and I'd love to know when it's updated.

**Comment by ditman** [comment]
@artificerchris certainly, check this PR: https://github.com/flutter/plugins/pull/3726 (still in progress, but already usable)

**Comment by artificerchris** [comment]
@ditman 

That didn't work out. I modified my pubspec like so:

```
google_maps_flutter:
    git: 
      url: https://github.com/ditman/plugins.git
      path: packages/google_maps_flutter/google_maps_flutter
  google_maps_flutter_web: 
    git: 
      url: https://github.com/ditman/plugins.git
      path: packages/google_maps_flutter/google_maps_flutter_web
```

and now I get

```
Error: Cannot run with sound null safety, because the following dependencies
don't support null safety:                                              
                                                                        
 - package:google_maps_flutter_web                                      
 - package:google_maps                                                  
 - package:js_wrapping      
 ```

**Comment by ditman** [comment]
@artificerchris don't forget the `ref: some-branch` git option. The branch in this case is: `google-maps-nnbd`; so:

```yaml
  google_maps_flutter_web: 
    git: 
      url: https://github.com/ditman/plugins.git
      ref: google-maps-nnbd
      path: packages/google_maps_flutter/google_maps_flutter_web
```

Also, I don't think I made any changes to `google_maps_flutter`; you don't need to override that one!

**Comment by artificerchris** [comment]
@ditman that did the trick! thank you! so now I just wait on the update on pub and swap back, right? I'm not super familiar with using a git version of a plugin.

**Comment by ditman** [comment]
> @ditman that did the trick! thank you! so now I just wait on the update on pub and swap back, right? I'm not super familiar with using a git version of a plugin.

@artificerchris correct! I'd recommend that you "Subscribe" to the PR above so you get notified of when it gets merged. We normally publish plugins a couple of hours after they've been merged into flutter/plugins `master`. Then you can just switch to the new version in pub.dev (which IIRC is going to be: `^0.3.0`)

**Comment by ditman** [comment]
@artificerchris you should be able to grab the null-safe google_maps_flutter_web plugin from pub now: `^0.3.0`.

More info: https://pub.dev/packages/google_maps_flutter_web/versions/0.3.0

**Comment by artificerchris** [comment]
@ditman Thanks for your work!

**Comment by d3roch4** [comment]
I am using version:
  google_maps_flutter: ^2.0.3
  google_maps_flutter_web: ^0.3.0
in channel master

```
// This work for me:
setState(() {
      marcadores.clear();
      marcadores.add(Marker(markerId: MarkerId(latLng.toString()), position: latLng));
    });
```

```
// This not work for me:
setState(() {
      marcadores.clear();
      marcadores.add(Marker(markerId: MarkerId("positionNow"), position: latLng));
    });
```

animateCamera not work any case:
>  mapCtrl?.animateCamera(CameraUpdate.newLatLngZoom(latLng, zoom));

**Comment by ditman** [comment]
@d3roch4 from what I've seen in [the example](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter/example/lib/place_marker.dart#L197-L207), you don't need to clear and add a new marker all the time, you should just update the with MarkerId that you want to move around.

> animateCamera not work any case

It should at least move to the new latLng and set the desired zoom level. In web, it seems the plugin decides if it's going to animate while moving the camera or not, depending on how far the new center is from the current one, but I haven't found documentation about this.

**Comment by celsoncustodio** [comment]
> @d3roch4 from what I've seen in [the example](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter/example/lib/place_marker.dart#L197-L207), you don't need to clear and add a new marker all the time, you should just update the Marker with MarkerId that you want to move around.
> 
> > animateCamera not work any case
> 
> It should at least move to the new latLng and set the desired zoom level. In web, it seems the plugin decides if it's going to animate while moving the camera or not, depending on how far the new center is from the current one, but I haven't found documentation about this.

Do you know any way to change marker position without clear all marker on flutter web ?

**Comment by ditman** [comment]
Digging deeper into this, I found this bit of the code that says:

> "Preserves the position from the [currentMarker], if set."
> [(Source)](https://github.com/flutter/plugins/blob/master/packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart#L219)

That is used to merge the properties of a Marker with whatever is already painted, HOWEVER, I don't really get why the position of the old marker matters (I blamed that line, and it's been there since the first iteration of the plugin, so maybe it was preventing an issue that doesn't happen anymore!).

Anyway, that seems like a bug that can be fixed. Knowing that, I now understand why Markers need to be deleted and re-created all the time. What people are achieving with that is to change the MarkerId to something that hasn't seen before by the Map widget (and that causes a repaint).

I'll try to hack around the issue above and post a quick example app in a few minutes.

**Comment by ditman** [comment]
Here's an example that seems to work:

It updates the position of the existing markers near the point where the map was clicked, and lets you add new markers when clicking on the Floating Action Button.

**main.dart**

```dart
import 'dart:math';

import 'package:flutter/material.dart';

import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:pointer_interceptor/pointer_interceptor.dart';

final rnd = Random();

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Moving a Marker Google Maps Flutter Web Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  // We use this value to paint new Markers close to the center of the last movement of the map.
  // Updates to this don't need to call setState, this is just a value that we care about some times.
  LatLng _center = LatLng(40.416775, -3.703790);

  // Will be updated when we tap on the map, or on the fAB.
  // Updates to this must call setState, so the GoogleMap gets re-painted.
  Map<MarkerId, Marker> _markers = {
    MarkerId('0'): Marker(markerId: MarkerId('madrid_initial'), position: LatLng(40.416775, -3.703790)),
  };

  // Will be initialized later by the [_onMapCreated] function.
  late GoogleMapController _controller;

  // This jitters a double a little bit, according to `scale`.
  // `scale` limits at which decimal point will the random function apply:
  // 0.00000001 -> tiny changes, 1.0 -> potentially large changes!
  double _jitterDouble(double val, double scale) {
    final min = val - (val * scale);
    final max = val + (val * scale);
    return rnd.nextDouble() * (max - min) + min;
  }

  // Jitters the lat/lng of a position, according to `scale`.
  LatLng _jitterPosition(LatLng position, double scale) {
    return LatLng(_jitterDouble(position.latitude, scale), _jitterDouble(position.longitude, scale));
  }

  // Creates a marker with a unique variation on a `MarkerId`, around a given `position`.
  Marker _timestampedMarker(MarkerId id, LatLng position) {
    return Marker(
      markerId: MarkerId('${id.value}_${DateTime.now().millisecondsSinceEpoch}'), 
      position: _jitterPosition(position, 0.002), // Jitter so not all markers fall in the same center...
    );
  }

  void _onMapCreated(GoogleMapController controller) {
    _controller = controller;
  }

  void _onCameraMove(CameraPosition position) {
    _center = position.target;
  }

  void _onMapTap(LatLng newPosition) {
    // For each marker, move it close to the latLng of the click...
    setState(() {
      _markers.forEach((markerId, marker) {
        _markers[markerId] = _timestampedMarker(markerId, newPosition);
        // When the Marker position bug is fixed, this can be replaced by:
        // _markers[markerId] = marker.copyWith(positionParam: _jitterPosition(newPosition));
      });
    });
    // Center the camera where the user clicked after a few milliseconds...
    Future.delayed(Duration(milliseconds: 150), () {
      _controller.animateCamera(CameraUpdate.newLatLng(newPosition));
    });
  }

  // Adds a marker near the `_center` of the map.
  void _addMarker() {
    setState(() {
      // Add a new marker to the set of markers available...
      final MarkerId id = MarkerId('${_markers.length}');
      _markers[id] = _timestampedMarker(id, _center);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
              child: GoogleMap(
                initialCameraPosition: CameraPosition(
                  target: _center,
                  zoom: 10,
                ), 
                zoomControlsEnabled: false,
                minMaxZoomPreference: MinMaxZoomPreference(10, 10),
                markers: _markers.values.toSet(),
                onMapCreated: _onMapCreated,
                onTap: _onMapTap,
                onCameraMove: _onCameraMove,
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: 
        // PointerInterceptor from package:pointer_interceptor, so clicks on the fAB
        // don't hit the underlying GoogleMap widget.
        PointerInterceptor(
          child: FloatingActionButton(
            onPressed: _addMarker,
            tooltip: 'Add Marker (near center)',
            child: Icon(Icons.add_location_alt_outlined),
          ),
        ),
    );
  }
}
```

**pubspec.yaml**

```yaml
name: gmaps_markers
description: A demo to demonstrate how to move a Marker with google maps web.
publish_to: none # Remove this line if you wish to publish to pub.dev

environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  cupertino_icons: ^1.0.2
  flutter:
    sdk: flutter
  google_maps_flutter: ^2.0.3
  google_maps_flutter_web: ^0.3.0 
  pointer_interceptor: ^0.9.0 # So the fAB doesn't cause a click on the Map.

dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:
  uses-material-design: true
```

**Comment by celsoncustodio** [comment]
You're the guy, thank you so much  @ditman .. It worked
